src/group_theory/quaternion_group.lean

```diff
@@ -0,0 +1,254 @@
+| (xa i) (a j) := xa (i + j)
+| (xa i) (xa j) := a (n + j - i)
+
+/--
+The identity `1` is given by `aⁱ`.
+-/
+private def one : quaternion_group n := a 0
+
+instance : inhabited (quaternion_group n) := ⟨one⟩
+
+/--
+The inverse of an element of the dihedral group.
+-/
+private def inv : quaternion_group n → quaternion_group n
+| (a i) := a (-i)
+| (xa i) := xa (n + i)
+
+/--
+The group structure on `quaternion_group n`.
+-/
+instance : group (quaternion_group n) :=
+{ mul := mul,
+  mul_assoc :=
+  begin
+    rintros (i | i) (j | j) (k | k);
+    simp only [mul];
+    abel,
+    simp only [neg_mul_eq_neg_mul_symm, one_mul, int.cast_one, gsmul_eq_mul, int.cast_neg,
+               add_right_inj],
+    calc -(n : zmod (2 * n)) = 0 - n : by rw zero_sub
+      ... = 2 * n - n : by { norm_cast, simp, }
+      ... = n : by ring
+  end,
+  one := one,
+  one_mul :=
+  begin
+    rintros (i | i),
+    exact congr_arg a (zero_add i),
+    exact congr_arg xa (sub_zero i),
+  end,
+  mul_one := begin
+    rintros (i | i),
+    exact congr_arg a (add_zero i),
+    exact congr_arg xa (add_zero i),
+  end,
+  inv := inv,
+  mul_left_inv := begin
+    rintros (i | i),
+    exact congr_arg a (neg_add_self i),
+    exact congr_arg a (sub_self (n + i)),
+  end }
+
+variable {n}
+
+@[simp] lemma a_mul_a (i j : zmod (2 * n)) : a i * a j = a (i + j) := rfl
+@[simp] lemma a_mul_xa (i j : zmod (2 * n)) : a i * xa j = xa (j - i) := rfl
+@[simp] lemma xa_mul_a (i j : zmod (2 * n)) : xa i * a j = xa (i + j) := rfl
+@[simp] lemma xa_mul_xa (i j : zmod (2 * n)) : xa i * xa j = a (n + j - i) := rfl
+
+lemma one_def : (1 : quaternion_group n) = a 0 := rfl
+
+private def fintype_helper : (zmod (2 * n) ⊕ zmod (2 * n)) ≃ quaternion_group n :=
+{ inv_fun := λ i, match i with
+                 | (a j) := sum.inl j
+                 | (xa j) := sum.inr j
+                 end,
+  to_fun := λ i, match i with
+                 | (sum.inl j) := a j
+                 | (sum.inr j) := xa j
+                 end,
+  left_inv := by rintro (x | x); refl,
+  right_inv := by rintro (x | x); refl }
+
+/-- The special case that more or less by definition `quaternion_group 0` is isomorphic to the
+infinite dihedral group. -/
+def quaternion_group_zero_equiv_dihedral_zero : quaternion_group 0 ≃* dihedral 0 :=
+{ to_fun := λ i, quaternion_group.rec_on i dihedral.r dihedral.sr,
+  inv_fun := λ i, match i with
+                | (dihedral.r j) := a j
+                | (dihedral.sr j) := xa j
+                end,
+  left_inv := by rintro (k | k); refl,
+  right_inv := by rintro (k | k); refl,
+  map_mul' := by { rintros (k | k) (l | l); { dsimp, simp, }, } }
+
+/-- Some of the lemmas on `zmod m` require that `m` is positive, as `m = 2 * n` is the case relevant
+in this file but we don't want to write `[fact (0 < 2 * n)]` we make this lemma a local instance. -/
+private lemma succ_mul_pos {m : ℕ} [fact (0 < n)] : fact (0 < (nat.succ m) * n) :=
+mul_pos (by norm_num) (_inst_1)
```

Comment:
```suggestion
private lemma succ_mul_pos {m : ℕ} [hn : fact (0 < n)] : fact (0 < (nat.succ m) * n) :=
mul_pos (by norm_num) hn
```
I don't know what mathlib style here would be. But not using auto-generated names is better imo.